### ISSUE TITLE
Error rendering report: Cannot read property 'startsWith' of undefined on safelySetHref function

### ISSUE THREAD

**Issue by derevandal** [GoogleChrome/lighthouse#12868]
### FAQ

- [X] Yes, my issue is not about [variability](https://github.com/GoogleChrome/lighthouse/blob/master/docs/variability.md) or [throttling](https://github.com/GoogleChrome/lighthouse/blob/master/docs/throttling.md).
- [X] Yes, my issue is not about a specific accessibility audit (file with [axe-core](https://github.com/dequelabs/axe-core) instead).

### URL

https://investmentu.com

### What happened?

I've set up an LHCI server with PSI auto collect.
But I'm getting one tiny issue related to `https://googlechrome.github.io/lighthouse/viewer/`.

When I click to open any report, the view complains:

```
Uncaught Error: Error rendering report: Cannot read property 'startsWith' of undefined
```

Debugging it, I "solved" this issue by setting the second parameter of this function to an empty string as default.

![image](https://user-images.githubusercontent.com/1340508/128359323-cef63f49-01c6-4a7e-9d4a-3a6a18489bf9.png)

https://user-images.githubusercontent.com/1340508/128362112-8f36fc61-22a6-48b2-a9c9-f28c179d2bd8.mp4

https://user-images.githubusercontent.com/1340508/128362114-a33d036f-8a5f-4229-8ad3-83b11f74adc2.mp4

### What did you expect?

The report should be shown properly.

### What have you tried?

I've used local overwrites to debug and found the solution.

### How were you running Lighthouse?

PageSpeed Insights

### Lighthouse Version

8.0.0

**Comment by connorjclark** [comment]
Thanks for the report, we're looking into it now.

In the future, to save yourself extra effort when reporting bugs, what is most important is the stack trace of the error, which I was able to find in your video. Sharing here for others:

![image](https://user-images.githubusercontent.com/4071474/128410601-a79d8613-8bd5-4591-82e2-f54b5672dfc2.png)

**Comment by connorjclark** [comment]
`details.url` is somehow undefined. The problematic detail value comes from `third-party-summary`:
```json
"entity": {
    "type": "link",
    "text": "PushCrew"
}
```

![image](https://user-images.githubusercontent.com/4071474/128411727-49c5a951-64e0-4ea2-924f-9f8e0d8cb1f9.png)

Looks good, right? Except remember: empty strings are dropped in LR because of proto shenanigans. https://github.com/GoogleChrome/lighthouse/blob/3ff45bb/core/lib/proto-preprocessor.js#L76

We should add a check in safelySetHref that the url is `string`. And do a full pass on all string values that are used to renderer stuff in the report renderer, making sure an empty/missing value won't explode things.

**Comment by connorjclark** [comment]
Reason why empty strings get dropped (not documented in our code so I asked shane): proto json encoder/decoder, by default, does not include fields if the value is the default for that type. This can be configured, but not on a per-type basis, so we must keep that option off.

> We should add a check in safelySetHref that the url is string. And do a full pass on all string values that are used to renderer stuff in the report renderer, making sure an empty/missing value won't explode things.

Alternatively: should we replace empty strings in LR with something like `'__LH__EMPTYSTRING__'` and then reset to `''` in renderer? would have to be on a breaking version.

**Comment by derevandal** [comment]
thank you, dude!
I appreciate your effort!

I open a PR that works for me: https://github.com/GoogleChrome/lighthouse/pull/12869

It just prevents `safelySetHref` to break hehe

maybe it can help

**Comment by paulirish** [comment]
@derevandal thanks for the PR! Agree that'll definitely fix it.   We will probably reach a bit deeper to try to fix the root issue, but we definitely appreciate you drafting that.


@connorjclark @brendankenny fixing the root issue feels like.. we address any of these situations. I think that's anything inside a proto `Struct`... which is just `Audit.Details`.  And then within there its strings of '', numbers of 0, booleans of false, etc.  
It does seem that we can allow our renderer to handle these dropped-defaults versions of LHR Audit.details or we do the bookmarked __LH__ replacement thing that connor mentioned. 
I bring up the larger scope issue in case we could catch some currently-silent bugs by tweaking our renderer types.

**Comment by brendankenny** [comment]
Yeah, it's definitely a bummer. Error handling actually came up in the [original review](https://github.com/GoogleChrome/lighthouse/pull/9067#discussion_r292701405) for `third-party-summary`, but we only considered bad URLs (like that empty string), not `undefined`. The try/catch happened to handle `undefined` as well so we never noticed, but there's no test for `undefined` (because why would there be), which would have caught the `url.startsWith('#')` issue introduced in #12796.

> Alternatively: should we replace empty strings in LR with something like `'__LH__EMPTYSTRING__'` and then reset to `''` in renderer? would have to be on a breaking version.

Converting all `string` LHR properties to `string|undefined` would be a major regression (and if we did it by hand for only the properties where this is possible we risk this happening again), so maybe we really should consider this option.

Maaaybe there's something in jspb in the years since we got this working where we have some wiggle room to make this better?

**Comment by brendankenny** [comment]
> I think that's anything inside a proto `Struct`... which is just `Audit.Details`. And then within there its strings of '', numbers of 0, booleans of false, etc.
> It does seem that we can allow our renderer to handle these dropped-defaults versions of LHR Audit.details

we could do this (we do it automatically to strip out `LH.IcuMessage`s and convert them all to strings, for instance), but that's almost every single value in audit details :/

**Comment by brendankenny** [comment]
for testing: [`proto-preprocessor` strips empty strings](https://github.com/GoogleChrome/lighthouse/blob/df9ca0cd3d23295459cf991953fae33ba4513bfb/lighthouse-core/lib/proto-preprocessor.js#L76-L96), so we could take one of our tests that checks a report renders `sample_v2` without error (is it just the viewer pptr test?) and run it with `processForProto(sample_v2)` as well (though we might need a different LHR with more empty strings, it looks like it wouldn't have hit this issue).

We could expand `processForProto`'s stripping of default values to numbers and booleans, too

**Comment by connorjclark** [comment]
> for testing ...

sample json doesn't include this particular case, and if there are others they are likely behind conditionals/fallbacks that we may not know to trigger. So it isn't full-proof, but it's still a good idea.

> Maaaybe there's something in jspb in the years since we got this working where we have some wiggle room to make this better?

I'll double-check if we really can't disable the default-exclusion behavior for only strings.

**Comment by brendankenny** [comment]
> I'll double-check if we really can't disable the default-exclusion behavior for only strings.

I guess it may not work for auditDetails in particular, since we have a lot of optional properties. I assume the JSON serializer won't be able to tell the difference between an optional string and a required empty string, since both will come as 0 bytes over the wire. That would be fine as long as we're doing falsy checks, but it's still broken on a type level.

**Comment by connorjclark** [comment]
No immediate plans to fix root cause.

We would like to:

1) consider what analytics we have that will let us know about these renderer errors when they happen
2) consider how to use our staging traffic stress test to try out new renderer at-scale before releasing to prod
